### PR TITLE
bpo-43510: Remove _pyio.OpenWrapper class

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue43510

### CODE REVIEW

File: Lib/_pyio.py [hunk 1]
```diff
@@ -323,8 +323,12 @@ class OpenWrapper:
...
     __doc__ = DocDescriptor()
 
-    def __new__(cls, *args, **kwargs):
-        return open(*args, **kwargs)
+    def __new__(cls, file, mode="r", buffering=-1, encoding=None,
+                *args, **kwargs):
+        # Emit the EncodingWarning for the caller.
+        if "b" not in mode:
+            encoding = text_encoding(encoding)
+        return open(file, mode, buffering, encoding, *args, **kwargs)
```
Comment (eryksun) [inline Lib/_pyio.py]: Are you checking this here instead of leaving it up to the `open` call in order to raise the warning at the right stack level?

Comment (methane) [reply Lib/_pyio.py]: Yes. `text_encoding()` is used for it.

Comment (eryksun) [reply Lib/_pyio.py]: Okay, so this is just a particular problem in an unusual case of a proxy call that needs to bring the `text_encoding` call forward, with special handling such as checking `if "b" not in mode`.

**Comment by vstinner** [comment]
OpenWrapper is really a weird beast. I don't know other functions which are defined this way. Can't we decorate ``def open`` with ``@staticmethod`` instead?

The io module doesn't use this hack: OpenWrapper is just an alias to open.

```
$ python3
Python 3.9.2 (default, Feb 20 2021, 00:00:00) 
>>> import io
>>> io.OpenWrapper
<built-in function open>
>>> io.OpenWrapper is io.open
True
```

**Comment by methane** [comment]
> OpenWrapper is really a weird beast. I don't know other functions which are defined this way. Can't we decorate `def open` with `@staticmethod` instead?

Nice idea.

**Comment by methane** [comment]
Oh, test is failing...

```
======================================================================
ERROR: test_read (test.test_univnewlines.PyTestCRLFNewlines)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython/Lib/test/test_univnewlines.py", line 51, in setUp
    with self.open(os_helper.TESTFN, self.WRITEMODE) as fp:
TypeError: 'staticmethod' object is not callable
```

**Comment by vstinner** [comment]
I opened a discussion on python-dev about OpenWrapper and staticmethod: https://mail.python.org/archives/list/python-dev@python.org/thread/QZ7SFW3IW3S2C5RMRJZOOUFSHHUINNME/

**Comment by vstinner** [comment]
> I opened a discussion on python-dev about OpenWrapper and staticmethod: https://mail.python.org/archives/list/python-dev@python.org/thread/QZ7SFW3IW3S2C5RMRJZOOUFSHHUINNME/

And I created https://bugs.python.org/issue43680 to remove undocumented io.OpenWrapper and _pyio.OpenWrapper.

**Comment by eryksun** [comment]
Maybe define `open` as a `classmethod` in a dummy class and set the bound method as global `open`. Or continue to define `open` as a function, and set `OpenWrapper` as a `functools.partial` instance, since it's builtin and doesn't insert a stack frame.

**Pull request closed without merging**